security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,423 @@
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            this._ek = null;
+            this._srk = null;
+            if (callback) {
+              if (err) {
+                callback(err);
+              } else {
+                callback(null, null);
+              }
+            }
+          },
+          connect: (connectCallback) => this._fsm.transition('connecting', connectCallback),
+          getEndorsementKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getEndorsementKey', callback);
+              }
+            });
+          },
+          getStorageRootKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getStorageRootKey', callback);
+              }
+            });
+          },
+          signWithIdentity: (dataToSign, callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('signWithIdentity', dataToSign, callback);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('activateSymmetricIdentity', identityKey, callback);
+              }
+            });
+          },
+          disconnect: (callback) => {
+            if (callback) {
+              callback();
+            }
+           }
+        },
+        connecting: {
+          _onEnter: (callback) => {
+            try {
+              let self = this;
+              this._tpm.connect(function (): void {
+                self._createPersistentPrimary({ name: 'EK', hierarchy: tss.Endorsement, handle: ekPersHandle, template: ekTemplate }, (ekCreateErr, ekPublicKey) => {
+                  if (ekCreateErr) {
+                    self._fsm.transition('disconnected', callback, ekCreateErr);
+                  } else {
+                    self._ek = ekPublicKey;
+                    self._createPersistentPrimary({ name: 'SRK', hierarchy: tss.Owner, handle: srkPersHandle, template: srkTemplate }, (srkCreateErr: Error, srkPublicKey: TPMT_PUBLIC) => {
+                      if (srkCreateErr) {
+                        self._fsm.transition('disconnected', callback, srkCreateErr);
+                      } else {
+                        self._srk = srkPublicKey;
+                        self._fsm.transition('connected', callback);
+                      }
+                    });
+                  }
+                });
+              });
+            } catch (err) {
```

Comment:
>catch (err) [](start = 14, length = 11)

Lame!   This only catches things thrown by the connect call, right?  In other words, if one of the callbacks above throws, this won't catch anything AFAIK.  It's too bad that _tpm.connect doesn't call it's callback with an error object like every single other function on the planet.

Reply:
Totally agree.  Will try to work with MSR to fix.

---
In reply to: [159309397](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r159309397) [](ancestors = 159309397)